x/genutil/genaccounts.go

```diff
@@ -15,133 +15,148 @@ import (
-// to be added for the account, `appendAcct` updates the account if already exists.
-// `vestingStart, vestingEnd and vestingAmtStr` respectively are the schedule start time, end time (unix epoch)
-// `moduleName` is the module name for which the account is being created
-// and coins to be appended to the account already in the genesis.json file.
-func AddGenesisAccount(
+type GenesisAccount struct {
+	// Base
+	Address string    `json:"address"`
+	Coins   sdk.Coins `json:"coins"`
+
+	// Vesting
+	VestingAmt   sdk.Coins `json:"vesting_amt,omitempty"`
+	VestingStart int64     `json:"vesting_start,omitempty"`
+	VestingEnd   int64     `json:"vesting_end,omitempty"`
+
+	// Module
+	ModuleName string `json:"module_name,omitempty"`
+}
+
+// AddGenesisAccounts adds genesis accounts to the genesis state.
+// Where `cdc` is the client codec, `addressCodec` is the address codec, `accounts` are the genesis accounts to add,
+// `appendAcct` updates the account if already exists, and `genesisFileURL` is the path/url of the current genesis file.
+func AddGenesisAccounts(
 	cdc codec.Codec,
 	addressCodec address.Codec,
-	accAddr sdk.AccAddress,
+	accounts []GenesisAccount,
 	appendAcct bool,
-	genesisFileURL, amountStr, vestingAmtStr string,
-	vestingStart, vestingEnd int64,
-	moduleName string,
+	genesisFileURL string,
 ) error {
-	addr, err := addressCodec.BytesToString(accAddr)
+	appState, appGenesis, err := genutiltypes.GenesisStateFromGenFile(genesisFileURL)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to unmarshal genesis state: %w", err)
 	}
 
-	coins, err := sdk.ParseCoinsNormalized(amountStr)
-	if err != nil {
-		return fmt.Errorf("failed to parse coins: %w", err)
-	}
+	authGenState := authtypes.GetGenesisStateFromAppState(cdc, appState)
+	bankGenState := banktypes.GetGenesisStateFromAppState(cdc, appState)
 
-	vestingAmt, err := sdk.ParseCoinsNormalized(vestingAmtStr)
+	accs, err := authtypes.UnpackAccounts(authGenState.Accounts)
 	if err != nil {
-		return fmt.Errorf("failed to parse vesting amount: %w", err)
+		return fmt.Errorf("failed to get accounts from any: %w", err)
 	}
 
-	// create concrete account type based on input parameters
-	var genAccount authtypes.GenesisAccount
+	newSupplyCoinsCache := sdk.NewCoins()
+	balanceCache := make(map[string]banktypes.Balance)
+	for _, acc := range accs {
+		for _, balance := range bankGenState.GetBalances() {
+			if balance.Address == acc.GetAddress().String() {
+				balanceCache[acc.GetAddress().String()] = balance
+			}
+		}
+	}
 
-	balances := banktypes.Balance{Address: addr, Coins: coins.Sort()}
-	baseAccount := authtypes.NewBaseAccount(accAddr, nil, 0, 0)
+	for _, acc := range accounts {
+		addr := acc.Address
+		coins := acc.Coins
 
-	if !vestingAmt.IsZero() {
-		baseVestingAccount, err := authvesting.NewBaseVestingAccount(baseAccount, vestingAmt.Sort(), vestingEnd)
+		accAddr, err := sdk.AccAddressFromBech32(addr)
```

Comment:
Should be using address codec

Reply:
3cc0ec323b774d5db8ad79babd6f6b8c8faadf33